src/Symfony/Bundle/FrameworkBundle/KernelBrowser.php

```diff
@@ -11,11 +11,27 @@
 
 namespace Symfony\Bundle\FrameworkBundle;
 
+use Symfony\Component\BrowserKit\Cookie;
+use Symfony\Component\Security\Core\Authentication\Token\UsernamePasswordToken;
+use Symfony\Component\Security\Core\User\UserInterface;
+
 /**
  * Client simulates a browser and makes requests to a Kernel object.
  *
  * @author Fabien Potencier <fabien@symfony.com>
  */
 class KernelBrowser extends Client
 {
+    public function loginUser(UserInterface $user, string $firewallContext = 'main'): self
+    {
+        $token = new UsernamePasswordToken($user, null, $firewallContext, $user->getRoles());
```

Comment:
Can we use a token specifically designed for testing here (perhaps an object as simple as possible)? The UsernamePasswordToken is specific to certain implementation. This also means that if you don't define any roles for your user, which is a very legit case, it would never consider you as authenticated, as this is a requirement for this particular token, but not Guard for example.

I've traced back the usages of this token in conjunction with "being authenticated", as it's injected into the session, and it _should_ accept any token, regardless of authenticator

Reply:
Thanks for the review! Sadly, my knowledge of the Symfony Security parts is not as good as yours ... so I don't understand your comment. I'm afraid that I'll need the full copy+pasteable code to update this pull request. Thanks!

Reply:
The issue is this line: https://github.com/symfony/symfony/blob/76cbdbff05a3eb12dcdd94f267b591ca09faae61/src/Symfony/Component/Security/Core/Authentication/Token/UsernamePasswordToken.php#L44

Meaning that whenever you want to authenticate, you are forced to create a user that has a minimum of 1 role, which isn't enforced in any other tokens. One option would be to use the `PostAuthenticationGuardToken`, the other would be to extends the `AbstractToken` and implement only what's not there yet.

There's also a `ConcreteToken` in `security/Core/Tests/Authentication/Token/AbstractTokenTest.php`, this is basically what I mean with a test specific token. Perhaps it's an idea to extract this token and make it a stub for multiple tests?